File: Go/task5.go
package main

import (
    "fmt"
)

// Вычисление цепной дроби для a/b
func computeContinuedFraction(a, b int) []int {
    coefficients := []int{}
    for b != 0 {
        quotient := a / b
        coefficients = append(coefficients, quotient)
        remainder := a % b
        a, b = b, remainder
    }
    return coefficients
}

// Вычисление НОД (алгоритм Евклида)
func computeGCD(a, b int) int {
    for b != 0 {
        a, b = b, a%b
    }
    return a
}

// Вывод компактного решения
func printCompactSolution(aPart, aCoeff, bPart, bCoeff int) {
    var k int
    if aCoeff != 0 && bCoeff != 0 {
        k1 := 0
        if aCoeff != 0 {
            k1 = aPart / aCoeff
        }
        k2 := 0
        if bCoeff != 0 {
            k2 = bPart / bCoeff
        }
        if k1 < k2 {
            k = k1
        } else {
            k = k2
        }
    }
    fmt.Printf("Одно компактное решение: a = %d, b = %d\n", aPart+aCoeff*k, bPart-bCoeff*k)
    fmt.Printf("Общее компактное решение: a = %d + %d*k, b = %d + %d*k, где k - целое число\n",
        aPart+aCoeff*k, aCoeff, bPart-bCoeff*k, bCoeff)
}

// Решение диофантова уравнения
func solveDiophantine(a, b, c int) {
    fraction := computeContinuedFraction(a, b)

    // Инициализация рекуррентных соотношений
    xCurrent, yCurrent := 0, 1
    xPrev, yPrev := 1, 0

    // Вычисление подходящих дробей
    for _, quotient := range fraction {
        xPrev, xCurrent = xCurrent, xPrev-quotient*xCurrent
        yPrev, yCurrent = yCurrent, yPrev-quotient*yCurrent
    }

    gcd := computeGCD(a, b)

    // Проверка существования решения
    if c%gcd != 0 {
        fmt.Println("Уравнение не имеет решений.")
        return
    }

    // Вычисление частного решения
    k := c / gcd
    fmt.Printf("Частное решение: x = %d, y = %d\n", xPrev*k, yPrev*k)

    // Вывод общего решения
    fmt.Printf("Общее решение: x = %d + %d*k, y = %d - %d*k, где k - целое число\n\n",
        xPrev*k, b/gcd, yPrev*k, a/gcd)

    // Вывод компактного решения
    printCompactSolution(xPrev*k, b/gcd, yPrev*k, a/gcd)
}

func main() {
    fmt.Println("Решение уравнения 1256x + 847y = 119\n")
    solveDiophantine(1256, 847, 119)
}

File: task5.cpp
#include <iostream>
#include <vector>

using namespace std;

// Вычисление цепной дроби для a/b
vector<int> ComputeContinuedFraction(int a, int b) {
    vector<int> coefficients;
    while (b != 0) {
        int quotient = a / b;
        coefficients.push_back(quotient);
        int remainder = a % b;
        a = b;
        b = remainder;
    }
    return coefficients;
}

// Алгоритм Евклида для нахождения НОД
int ComputeGCD(int a, int b) {
    while (b != 0) {
        int temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

// Вывод компактного решения
void PrintCompactSolution(int a_part, int a_coeff, int b_part, int b_coeff) {
    int k = (a_part / a_coeff < b_part / b_coeff) ? 
             a_part / a_coeff : b_part / b_coeff;

    cout << "Одно компактное решение: a = " << a_part + (a_coeff * k) 
         << ", b = " << b_part + (-b_coeff * k) << endl;

    cout << "Общее компактное решение: a = " << a_part + (a_coeff * k) << " + " 
         << a_coeff << " * k, b = " << b_part + (-b_coeff * k) << " + " 
         << b_coeff << " * k, где k - целое число" << endl;
}

// Решение диофантова уравнения ax + by = c
void SolveDiophantine(int a, int b, int c) {
    vector<int> fraction = ComputeContinuedFraction(a, b);
    
    // Инициализация рекуррентных соотношений
    int x_current = 0, y_current = 1;
    int x_prev = 1, y_prev = 0;
    
    // Вычисление подходящих дробей
    for (int quotient : fraction) {
        int x_temp = x_current;
        x_current = x_prev - quotient * x_current;
        x_prev = x_temp;
        
        int y_temp = y_current;
        y_current = y_prev - quotient * y_current;
        y_prev = y_temp;
    }

    int gcd = ComputeGCD(a, b);
    
    // Проверка существования решения
    if (c % gcd != 0) {
        cout << "Уравнение не имеет решений." << endl;
        return;
    }
    
    // Вычисление частного решения
    int k = c / gcd;
    cout << "Частное решение: x = " << x_prev * k << ", y = " << y_prev * k << endl;
    
    // Вывод общего решения
    cout << "Общее решение: x = " << x_prev * k << " + " << b/gcd << " * k, "
         << "y = " << y_prev * k << " - " << a/gcd << " * k, где k - целое число" << endl << endl;

    // Вывод компактного решения
    PrintCompactSolution(x_prev * k, b/gcd, y_prev * k, a/gcd);
}

int main() {
    cout << "Решение уравнения 1256x + 847y = 119" << endl << endl;
    SolveDiophantine(1256, 847, 119);

    return 0;
}